src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1412,44 +1421,97 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
-                    if (!ArrayUtils.IsEnumerable(p) || p == typeof(string))
-                    {
-                        // arg is an enumerable type, param is not.
-                        return false;
-                    }
-                    var argRank = GetArgumentRank(exp);
-                    if (argRank == -1)
+                    if (exp.ReplicationGuides.Count > 0) return false;
+                    Type t;
+                    if (!astTypeInfoMap.TryGetValue(exp.ID, out t))
                     {
-                        // non-rectangular (jagged) array best handled by replication
-                        return false;
+                        throw new Exception("unkown ast type");
                     }
-                    if (argRank != GetRank(p)) return false;
+                    if (!DoesParamRankMatchArg(fep, p, t, argIndex)) return false;
                 }
-                else if (args[argIndex] is IdentifierNode idn)
+                else if (currentArg is IdentifierNode idn)
                 {
                     if (idn.ReplicationGuides.Count > 0) return false;
 
                     var t = variables[idn.Value].Item2;
                     if (t == typeof(object)) return false;
 
-                    if (!p.Equals(t))
+                    if(!DoesParamRankMatchArg(fep,p, t,argIndex)) return false;
+                }
+                else if (currentArg is RangeExprNode rgnNode)
+                {
+                    Type t;
+                    if (!astTypeInfoMap.TryGetValue(rgnNode.ID, out t))
                     {
-                        var argRank = GetRank(t);
-                        var paramRank = GetRank(p);
-                        if (argRank != paramRank) return false;
-
-                        // If both have rank 0, it could also be because their type info is ambiguous.
-                        if (argRank == 0 && paramRank == 0) return false;
+                        throw new Exception("unkown ast type");
                     }
+                    if (!DoesParamRankMatchArg(fep,p, t,argIndex)) return false;
                 }
                 else return false;
                 argIndex++;
             }
             return true;
         }
 
+        /// <summary>
+        /// Check that a parameter's rank and type are compatible with the given argument.
+        /// This function besides checking rank directly, also does checks using the FEP attributes.
+        /// </summary>
+        /// <param name="fep">the function end point wrapper</param>
+        /// <param name="paramType">type of param on the fep</param>
+        /// <param name="argType">actual argument type</param>
+        /// <param name="paramIndex">parameter index in list of params to fep.</param>
+        /// <returns></returns>
+        private bool DoesParamRankMatchArg(ProtoCore.CLRFunctionEndPoint fep,Type paramType, Type argType, int paramIndex)
+        {
+            if (!paramType.Equals(argType))
```

Comment:
Minor but it would've been easier to read and less indentation if you inverted this `if`:
```
if(paramType.Equals(argType)) return true;
...
```